toolchain/install/installation.cpp

```diff
@@ -0,0 +1,98 @@
+
+#include "common/check.h"
+#include "common/vlog.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/Path.h"
+
+namespace Carbon {
+
+Installation::Installation(llvm::StringRef exe_path,
+                           llvm::raw_ostream* vlog_stream)
+    : vlog_stream_(vlog_stream) {
+  // Map from the executable path from the executable path to an installation
+  // prefix path.
+
+  // TODO: Detect a Windows executable path and use custom logic to map to the
+  // correct install prefix for that platform.
+
+  // First, see if the binary is in a `bin` directory. If so, remove it
+  // assuming this is a FHS-like install prefix. We start with this pattern both
+  // because we expect it to be the "production" pattern and so useful to
+  // involve a minimal number of filesystem accesses, and because even in
+  // testing or development builds, when we _can_ use this pattern, we prefer it
+  // to match the behavior of genuine installations as closely as possible.
+  install_prefix_.assign(exe_path);
+  llvm::sys::path::remove_filename(install_prefix_);
+  if (auto rbegin_it = llvm::sys::path::rbegin(install_prefix_),
+      rend_it = llvm::sys::path::rend(install_prefix_);
+      rbegin_it != rend_it) {
+    // Step over a trailing `.` signifying the current directory. Whether this
+    // occurs for paths is ambiguous between the documentation and the observed
+    // behavior, so this code works to be resilient.
+    if (*rbegin_it == ".") {
+      ++rbegin_it;
+    }
+    if (rbegin_it != rend_it && *rbegin_it == "bin") {
+      // Rather than trying to remove the `bin` directory, walk upwards by
+      // appending `..`. We also assume POSIX style at this point.
+      llvm::sys::path::append(install_prefix_, llvm::sys::path::Style::posix,
+                              "../");
+      return;
+    }
+  }
+
+  // Next, check for a Bazel runfiles tree. This will match for development
+  // builds and direct execution of tests or other just-built binaries.
+  install_prefix_.assign(exe_path);
+  install_prefix_.append(".runfiles");
+  if (llvm::sys::fs::is_directory(install_prefix_)) {
+    llvm::sys::path::append(install_prefix_, llvm::sys::path::Style::posix,
+                            "_main/toolchain/install/prefix_root/");
+    CARBON_CHECK(llvm::sys::fs::is_directory(install_prefix_))
+        << "Found runfiles tree but not an install prefix directory!";
+    return;
+  }
+
+  // Lastly, check for a test runtimes tree environment variable. This will
+  // match when running under Bazel and thus inside the runfiles tree. We check
+  // this last of all in part because that allows testing of all of the prior
+  // patterns.
+  install_prefix_.assign(llvm::StringRef(getenv("TEST_SRCDIR")));
+  if (llvm::sys::fs::is_directory(install_prefix_)) {
+    llvm::sys::path::append(install_prefix_, llvm::sys::path::Style::posix,
+                            "_main/toolchain/install/prefix_root/");
```

Comment:
I really didn't want to have both of these either. =/

Sadly, the way Bazel executes test makes it impossible as far as I can tell to use the same logic for the following three things:

1) Running a unittest.
2) Running inside of a `genrule` or as part of a build step.
3) Running the binary directly out of `./bazel-bin/...`.

I'm not 100% certain whether `bazel run` behaves like (2) or (3) here, but it is one of them.

From what I can tell, (1) has `TEST_SRCDIR` and `RUNFILES_DIR` set, (2) has `RUNFILES_DIR` and `binary_name.runfiles/` symlink, and (3) obviously only has `binary_name.runfiles/`. For some reason (1) doesn't have `binary_name.runfiles/`, I really don't know why.

Anyways, that's how I ended up with the logic breakdown here. I really wanted to make sure that directly running `./bazel-bin/...` works, and that seemed to handle everything but tests, so I added `TEST_SRCDIR` for that case. I couldn't find a single set of logic that would handle all of these, but if I'm missing something, that would be fantastic.

Regarding `GetRunfiles`, as you mention, the license is a worry. And while it is running as *part* of the test, the actual detection code is in the "production" binary, so I don't see a way to avoid it. The API also seems quite tricky to use here, although there is probably a way. WDYT, any better ideas?

Reply:
I was thinking for `GetRunfiles` you could refactor to allow the test to pass in the path. Note that not all test environments put runfiles in the structure you're expecting here.

Reply:
Looking closer at the implementation... On Runfiles, EnvVars() will contain an entry whose key is RUNFILES_DIR which could be used. This looks like it's independent of how it's found, and could probably be used instead of RLocation.

Note, I think TEST_SRCDIR might be used in CreateForTest, but isn't necessary, and isn't used at all in some alternate implementations of this logic.